## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#aaae193c-3e48-4749-92ae-bcbf0b46ff5f) |
<a name="aaae193c-3e48-4749-92ae-bcbf0b46ff5f"></a>
### 11251-56896d0/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:32282
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8f27014d844e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 10 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 2 hours 15 minutes and 32 seconds
{
  "name": "aws-sdk-php",
  "function": "createMultipartUpload ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5ff88ad7068e2",
    "Object": "obj1"
  },
  "duration": "8039762",
  "status": "FAIL",
  "error": "Error executing \"UploadPart\" on \"http://minio-c2.minio.io:32282/aws-sdk-php-5ff88ad7068e2/obj1?uploadId=2ac6fcbcd19bbbf4&partNumber=1\"; AWS HTTP error: Client error: `PUT http://minio-c2.minio.io:32282/aws-sdk-php-5ff88ad7068e2/obj1?uploadId=2ac6fcbcd19bbbf4&partNumber=1` resulted in a `409 Conflict` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>BlockCountExceedsLimit</Code><Message>-&gt; github.com/Azure/azure-s (truncated...)\n BlockCountExceedsLimit (client): -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====\nDescription=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\nRequestId:9b4b6e69-001e-0051-05ef-e5f31b000000\nTime:2021-01-08T18:55:22.8860395Z, Details: \n   Code: BlockCountExceedsLimit\n   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5ff88ad7068e2/obj1?blockid=YzIwZjBhOTgtNGIyNS00NzJhLThjMzgtOTEzNmJkNjUyZmNk&comp=block&timeout=61\n   Authorization: REDACTED\n   Content-Length: [25]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]\n   X-Ms-Client-Request-Id: [d18a18c7-bfd5-4222-62a9-fe80ed2b341c]\n   X-Ms-Date: [Fri, 08 Jan 2021 18:55:22 GMT]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\n   Content-Length: [269]\n   Content-Type: [application/xml]\n   Date: [Fri, 08 Jan 2021 18:55:22 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Client-Request-Id: [d18a18c7-bfd5-4222-62a9-fe80ed2b341c]\n   X-Ms-Error-Code: [BlockCountExceedsLimit]\n   X-Ms-Request-Id: [9b4b6e69-001e-0051-05ef-e5f31b000000]\n   X-Ms-Version: [2019-02-02]\n\n\n - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>BlockCountExceedsLimit</Code><Message>-&gt; github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42&#xA;===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====&#xA;Description=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.&#xA;RequestId:9b4b6e69-001e-0051-05ef-e5f31b000000&#xA;Time:2021-01-08T18:55:22.8860395Z, Details: &#xA;   Code: BlockCountExceedsLimit&#xA;   PUT https://minioazure.blob.core.windows.net/aws-sdk-php-5ff88ad7068e2/obj1?blockid=YzIwZjBhOTgtNGIyNS00NzJhLThjMzgtOTEzNmJkNjUyZmNk&amp;comp=block&amp;timeout=61&#xA;   Authorization: REDACTED&#xA;   Content-Length: [25]&#xA;   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]&#xA;   X-Ms-Client-Request-Id: [d18a18c7-bfd5-4222-62a9-fe80ed2b341c]&#xA;   X-Ms-Date: [Fri, 08 Jan 2021 18:55:22 GMT]&#xA;   X-Ms-Version: [2019-02-02]&#xA;   --------------------------------------------------------------------------------&#xA;   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.&#xA;   Content-Length: [269]&#xA;   Content-Type: [application/xml]&#xA;   Date: [Fri, 08 Jan 2021 18:55:22 GMT]&#xA;   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]&#xA;   X-Ms-Client-Request-Id: [d18a18c7-bfd5-4222-62a9-fe80ed2b341c]&#xA;   X-Ms-Error-Code: [BlockCountExceedsLimit]&#xA;   X-Ms-Request-Id: [9b4b6e69-001e-0051-05ef-e5f31b000000]&#xA;   X-Ms-Version: [2019-02-02]&#xA;&#xA;&#xA;</Message><Key>obj1</Key><BucketName>aws-sdk-php-5ff88ad7068e2</BucketName><Resource>/aws-sdk-php-5ff88ad7068e2/obj1</Resource><RequestId>16584F52ACE6E801</RequestId><HostId>f71e5a98-df00-409e-8a74-ed768167a8e0</HostId></Error>"
}
(3/15) Running aws-sdk-ruby tests ... done in 24 seconds
(4/15) Running awscli tests ... FAILED in 5 minutes and 42 seconds
{
  "name": "awscli",
  "duration": 313184,
  "function": "aws --endpoint-url http://minio-c2.minio.io:32282 s3api upload-part --bucket awscli-mint-test-bucket-24791 --key awscli-mint-test-bucket-24791-object --body /mint/data/datafile-5-MB --upload-id 8dec4a15a7704964 --part-number 1\n",
  "status": "FAIL",
  "error": ""
}
(4/15) Running healthcheck tests ... done in 1 seconds
(5/15) Running mc tests ... FAILED in 1 hours 18 minutes and 53 seconds
{
  "name": "mc",
  "duration": "4674033",
  "function": "test_mirror_list_objects",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-22398 --quiet --no-color mirror /mint/data myminio/mc-test-bucket-13410  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-13410/datafile-1-MB`\n`/mint/data/datafile-0-b` -> `myminio/mc-test-bucket-13410/datafile-0-b`\n`/mint/data/datafile-1-kB` -> `myminio/mc-test-bucket-13410/datafile-1-kB`\n`/mint/data/datafile-1-b` -> `myminio/mc-test-bucket-13410/datafile-1-b`\n`/mint/data/datafile-1.03-MB` -> `myminio/mc-test-bucket-13410/datafile-1.03-MB`\n`/mint/data/datafile-100-kB` -> `myminio/mc-test-bucket-13410/datafile-100-kB`\n`/mint/data/datafile-129-MB` -> `myminio/mc-test-bucket-13410/datafile-129-MB`\n`/mint/data/datafile-10-kB` -> `myminio/mc-test-bucket-13410/datafile-10-kB`\n`/mint/data/datafile-10-MB` -> `myminio/mc-test-bucket-13410/datafile-10-MB`\n`/mint/data/datafile-11-MB` -> `myminio/mc-test-bucket-13410/datafile-11-MB`\n`/mint/data/datafile-6-MB` -> `myminio/mc-test-bucket-13410/datafile-6-MB`\n`/mint/data/datafile-5243880-b` -> `myminio/mc-test-bucket-13410/datafile-5243880-b`\n`/mint/data/datafile-5-MB` -> `myminio/mc-test-bucket-13410/datafile-5-MB`\n`/mint/data/datafile-33-kB` -> `myminio/mc-test-bucket-13410/datafile-33-kB`\n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-13410/datafile-65-MB`\nmc: <ERROR> Failed to copy `/mint/data/datafile-129-MB`. -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====\nDescription=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\nRequestId:d458e33f-301e-007b-61fb-e5865e000000\nTime:2021-01-08T20:20:21.5223836Z, Details: \n   Code: BlockCountExceedsLimit\n   PUT https://minioazure.blob.core.windows.net/mc-test-bucket-13410/datafile-129-MB?blockid=NDhiNTJjN2YtZjE2NS00YzIxLWJmZTctMzEyMGRkZDMxY2Q1&comp=block&timeout=61\n   Authorization: REDACTED\n   Content-Length: [25]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]\n   X-Ms-Client-Request-Id: [b3de548a-1644-4a2f-45b3-4716e374e9a6]\n   X-Ms-Date: [Fri, 08 Jan 2021 20:20:21 GMT]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\n   Content-Length: [269]\n   Content-Type: [application/xml]\n   Date: [Fri, 08 Jan 2021 20:20:20 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Client-Request-Id: [b3de548a-1644-4a2f-45b3-4716e374e9a6]\n   X-Ms-Error-Code: [BlockCountExceedsLimit]\n   X-Ms-Request-Id: [d458e33f-301e-007b-61fb-e5865e000000]\n   X-Ms-Version: [2019-02-02]\n\n\n\nTotal: 0 B, Transferred: 341.60 MiB, Speed: 74.84 KiB/s"
}
(5/15) Running minio-dotnet tests ... done in 5 minutes and 18 seconds
(6/15) Running minio-go tests ... FAILED in 10 minutes and 23 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
      "Object": "dstObject",
      "Encryption": null,
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "sourceList": [
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": false,
        "Start": 0,
        "End": 0,
        "Encryption": null
      },
      {
        "Bucket": "minio-go-test-b0h1t2z50d0hgltg",
        "Object": "srcObject",
        "VersionID": "",
        "MatchETag": "",
        "NoMatchETag": "",
        "MatchModifiedSince": "0001-01-01T00:00:00Z",
        "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
        "MatchRange": true,
        "Start": 0,
        "End": 0,
        "Encryption": null
      }
    ]
  },
  "duration": 603029,
  "error": "Put \"http://minio-c2.minio.io:32282/minio-go-test-b0h1t2z50d0hgltg/dstObject?partNumber=1&uploadId=2eb3e83385c2c6f3\": net/http: timeout awaiting response headers",
  "function": "ComposeObject(destination, sourceList)",
  "message": "ComposeObject failed",
  "name": "minio-go: testComposeMultipleSources",
  "status": "FAIL"
}
(6/15) Running minio-java tests ... FAILED in 13 minutes and 44 seconds
{
  "name": "minio-java",
  "function": "putObject()",
  "args": "[multi-part upload]",
  "duration": 300609,
  "status": "FAIL",
  "error": "java.net.SocketTimeoutException: timeout >>> [okio.SocketAsyncTimeout.newTimeoutException(JvmOkio.kt:143), okio.AsyncTimeout.access$newTimeoutException(AsyncTimeout.kt:162), okio.AsyncTimeout$sink$1.write(AsyncTimeout.kt:340), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235), okio.RealBufferedSink.write(RealBufferedSink.kt:145), okhttp3.internal.http1.Http1ExchangeCodec$KnownLengthSink.write(Http1ExchangeCodec.kt:271), okio.ForwardingSink.write(ForwardingSink.kt:29), okhttp3.internal.connection.Exchange$RequestBodySink.write(Exchange.kt:218), okio.RealBufferedSink.emitCompleteSegments(RealBufferedSink.kt:235), okio.RealBufferedSink.write(RealBufferedSink.kt:193), io.minio.HttpRequestBody.writeTo(HttpRequestBody.java:78), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:59), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109), okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201), okhttp3.internal.connection.RealCall.execute(RealCall.kt:154), io.minio.MinioClient.execute(MinioClient.java:647), io.minio.MinioClient.uploadPart(MinioClient.java:4747), io.minio.MinioClient.putObject(MinioClient.java:2738), io.minio.MinioClient.putObject(MinioClient.java:2830), FunctionalTest.testPutObject(FunctionalTest.java:763), FunctionalTest.putObject(FunctionalTest.java:837), FunctionalTest.runObjectTests(FunctionalTest.java:3751), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(6/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 81,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:79:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (internal/streams/readable.js:684:10) at endReadableNT (internal/streams/readable.js:1327:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(6/15) Running minio-py tests ... FAILED in 46 minutes and 17 seconds
{
  "name": "minio-py:test_fput_object_large_file",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17",
    "object_name": "2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large",
    "file_path": "/mint/data/datafile-11-MB",
    "metadata": {
      "x-amz-storage-class": "STANDARD_IA"
    }
  },
  "message": "S3 operation failed; code: BlockCountExceedsLimit, message: -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====\nDescription=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\nRequestId:1ec31ff9-301e-00f3-1f06-e63e87000000\nTime:2021-01-08T21:36:48.9578808Z, Details: \n   Code: BlockCountExceedsLimit\n   PUT https://minioazure.blob.core.windows.net/minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17/2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large?blockid=OThiYTFiNmYtZGRmYi00NGUzLTkyNTAtZmIzOGRlYWQyMzBi&comp=block&timeout=61\n   Authorization: REDACTED\n   Content-Length: [25]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]\n   X-Ms-Client-Request-Id: [0282a013-4a7d-4515-6d1b-f71e0d5ccdf7]\n   X-Ms-Date: [Fri, 08 Jan 2021 21:36:48 GMT]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\n   Content-Length: [269]\n   Content-Type: [application/xml]\n   Date: [Fri, 08 Jan 2021 21:36:48 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Client-Request-Id: [0282a013-4a7d-4515-6d1b-f71e0d5ccdf7]\n   X-Ms-Error-Code: [BlockCountExceedsLimit]\n   X-Ms-Request-Id: [1ec31ff9-301e-00f3-1f06-e63e87000000]\n   X-Ms-Version: [2019-02-02]\n\n\n, resource: /minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17/2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large, request_id: 16585CEC7A4479BA, host_id: f71e5a98-df00-409e-8a74-ed768167a8e0, bucket_name: minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17, object_name: 2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 356, in test_fput_object_large_file\n    _test_fput_object(bucket_name, object_name, _LARGE_FILE, metadata, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 308, in _test_fput_object\n    _CLIENT.fput_object(bucket_name, object_name, filename, sse=sse)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1010, in fput_object\n    tags=tags, retention=retention, legal_hold=legal_hold,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1792, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1770, in put_object\n    result = pool.result()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 722, in result\n    raise self._exceptions_queue.get()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/helpers.py\", line 674, in run\n    result = func(*args, **kargs)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1634, in _upload_part_task\n    return args[5], self._upload_part(*args)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1628, in _upload_part\n    bucket_name, object_name, data, headers, query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1610, in _put_object\n    query_params=query_params,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BlockCountExceedsLimit, message: -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====\nDescription=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\nRequestId:1ec31ff9-301e-00f3-1f06-e63e87000000\nTime:2021-01-08T21:36:48.9578808Z, Details: \n   Code: BlockCountExceedsLimit\n   PUT https://minioazure.blob.core.windows.net/minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17/2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large?blockid=OThiYTFiNmYtZGRmYi00NGUzLTkyNTAtZmIzOGRlYWQyMzBi&comp=block&timeout=61\n   Authorization: REDACTED\n   Content-Length: [25]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]\n   X-Ms-Client-Request-Id: [0282a013-4a7d-4515-6d1b-f71e0d5ccdf7]\n   X-Ms-Date: [Fri, 08 Jan 2021 21:36:48 GMT]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\n   Content-Length: [269]\n   Content-Type: [application/xml]\n   Date: [Fri, 08 Jan 2021 21:36:48 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Client-Request-Id: [0282a013-4a7d-4515-6d1b-f71e0d5ccdf7]\n   X-Ms-Error-Code: [BlockCountExceedsLimit]\n   X-Ms-Request-Id: [1ec31ff9-301e-00f3-1f06-e63e87000000]\n   X-Ms-Version: [2019-02-02]\n\n\n, resource: /minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17/2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large, request_id: 16585CEC7A4479BA, host_id: f71e5a98-df00-409e-8a74-ed768167a8e0, bucket_name: minio-py-test-a07cb5f9-c9fb-4a4a-8f85-f27d5956bd17, object_name: 2ebfb06c-e9ee-46c9-8120-65fd8cf0174a-large\n",
  "duration": 2773424
}
(6/15) Running s3cmd tests ... FAILED in 2 hours 17 minutes and 24 seconds
{
  "name": "s3cmd",
  "duration": "8195113",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "ERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-10457/s3cmd-test-object-12473 6f18d81e77023973\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 6f18d81e77023973 put ...\nto continue the upload.\nERROR: S3 error: 409 (BlockCountExceedsLimit): -> github.com/Azure/azure-storage-blob-go/azblob.newStorageError, github.com/Azure/azure-storage-blob-go@v0.10.0/azblob/zc_storage_error.go:42\n===== RESPONSE ERROR (ServiceCode=BlockCountExceedsLimit) =====\nDescription=The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\nRequestId:5936951f-401e-0116-4519-e66a25000000\nTime:2021-01-08T23:54:14.2138969Z, Details: \n   Code: BlockCountExceedsLimit\n   PUT https://minioazure.blob.core.windows.net/s3cmd-test-bucket-10457/s3cmd-test-object-12473?blockid=OGNhYTgyZTItNDlhMS00NGYzLWJmNTktYWQ3OTI4MmMzYTc3&comp=block&timeout=61\n   Authorization: REDACTED\n   Content-Length: [25]\n   User-Agent: [APN/1.0 MinIO/1.0 MinIO/2021-01-08T16:33:19Z]\n   X-Ms-Client-Request-Id: [4d1eaa90-48c4-4917-472c-ef76cfd1afb6]\n   X-Ms-Date: [Fri, 08 Jan 2021 23:54:14 GMT]\n   X-Ms-Version: [2019-02-02]\n   --------------------------------------------------------------------------------\n   RESPONSE Status: 409 The uncommitted block count cannot exceed the maximum limit of 100,000 blocks.\n   Content-Length: [269]\n   Content-Type: [application/xml]\n   Date: [Fri, 08 Jan 2021 23:54:13 GMT]\n   Server: [Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0]\n   X-Ms-Client-Request-Id: [4d1eaa90-48c4-4917-472c-ef76cfd1afb6]\n   X-Ms-Error-Code: [BlockCountExceedsLimit]\n   X-Ms-Request-Id: [5936951f-401e-0116-4519-e66a25000000]\n   X-Ms-Version: [2019-02-02]"
}
(6/15) Running s3select tests ... done in 42 seconds
(7/15) Running security tests ... done in 0 seconds

Executed 7 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/gateway/azure/gateway-azure.go
+++ b/cmd/gateway/azure/gateway-azure.go
@@ -51,30 +51,19 @@ import (
 	minio "github.com/minio/minio/cmd"
 )
 
-var (
-	azureUploadChunkSize   = getUploadChunkSizeFromEnv(azureChunkSizeEnvVar, strconv.Itoa(azureDefaultUploadChunkSize/humanize.MiByte))
-	azureSdkTimeout        = time.Duration(azureUploadChunkSize/humanize.MiByte) * azureSdkTimeoutPerMb
-	azureUploadConcurrency = azureUploadMaxMemoryUsage / azureUploadChunkSize
+const (
+	azureDefaultUploadChunkSizeMB = 25
+	azureDownloadRetryAttempts    = 5
+	azureS3MinPartSize            = 5 * humanize.MiByte
+	metadataObjectNameTemplate    = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x/azure.json"
+	azureMarkerPrefix             = "{minio}"
+	metadataPartNamePrefix        = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x"
+	maxPartsCount                 = 10000
 )
 
-const (
-	// The defaultDialTimeout for communicating with the cloud backends is set
-	// to 30 seconds in utils.go; the Azure SDK recommends to set a timeout of 60
-	// seconds per MB of data a client expects to upload so we must transfer less
-	// than 0.5 MB per chunk to stay within the defaultDialTimeout tolerance.
-	// See https://github.com/Azure/azure-storage-blob-go/blob/fc70003/azblob/zc_policy_retry.go#L39-L44 for more details.
-	// To change the upload chunk size, set the environmental variable MINIO_AZURE_CHUNK_SIZE_MB with a (float) value between 0 and 100
-	azureDefaultUploadChunkSize = 25 * humanize.MiByte
-	azureSdkTimeoutPerMb        = 60 * time.Second
-	azureUploadMaxMemoryUsage   = 100 * humanize.MiByte
-	azureChunkSizeEnvVar        = "MINIO_AZURE_CHUNK_SIZE_MB"
-
-	azureDownloadRetryAttempts = 5
-	azureS3MinPartSize         = 5 * humanize.MiByte
-	metadataObjectNameTemplate = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x/azure.json"
-	azureMarkerPrefix          = "{minio}"
-	metadataPartNamePrefix     = minio.GatewayMinioSysTmp + "multipart/v1/%s.%x"
-	maxPartsCount              = 10000
+var (
+	azureUploadChunkSize   int
+	azureUploadConcurrency int
 )
 
 func init() {
@@ -94,7 +83,6 @@ EXAMPLES:
   1. Start minio gateway server for Azure Blob Storage backend on custom endpoint.
      {{.Prompt}} {{.EnvVarSetCommand}} MINIO_ROOT_USER{{.AssignmentOperator}}azureaccountname
      {{.Prompt}} {{.EnvVarSetCommand}} MINIO_ROOT_PASSWORD{{.AssignmentOperator}}azureaccountkey
-     {{.Prompt}} {{.EnvVarSetCommand}} MINIO_AZURE_CHUNK_SIZE_MB {{.AssignmentOperator}}0.25
      {{.Prompt}} {{.HelpName}} https://azureaccountname.blob.custom.azure.endpoint
 
   2. Start minio gateway server for Azure Blob Storage backend with edge caching enabled.
@@ -106,8 +94,8 @@ EXAMPLES:
      {{.Prompt}} {{.EnvVarSetCommand}} MINIO_CACHE_AFTER{{.AssignmentOperator}}3
      {{.Prompt}} {{.EnvVarSetCommand}} MINIO_CACHE_WATERMARK_LOW{{.AssignmentOperator}}75
      {{.Prompt}} {{.EnvVarSetCommand}} MINIO_CACHE_WATERMARK_HIGH{{.AssignmentOperator}}85
-     {{.Prompt}} {{.EnvVarSetCommand}} MINIO_AZURE_CHUNK_SIZE_MB {{.AssignmentOperator}}25
      {{.Prompt}} {{.HelpName}}
+
 `
 
 	minio.RegisterGatewayCommand(cli.Command{
@@ -140,27 +128,6 @@ func azureGatewayMain(ctx *cli.Context) {
 	minio.StartGateway(ctx, &Azure{host})
 }
 
-// getUploadChunkSizeFromEnv returns the parsed chunk size from the environmental variable 'MINIO_AZURE_CHUNK_SIZE_MB'
-// The environmental variable should be a floating point number between 0 and 100 representing the MegaBytes
-// The returned value is an int representing the size in bytes
-func getUploadChunkSizeFromEnv(envvar string, defaultValue string) int {
-	envChunkSize := env.Get(envvar, defaultValue)
-
-	i, err := strconv.ParseFloat(envChunkSize, 64)
-	if err != nil {
-		logger.LogIf(context.Background(), err)
-		return azureDefaultUploadChunkSize
-	}
-
-	if i <= 0 || i > 100 {
-		logger.LogIf(context.Background(), fmt.Errorf("ENV '%v' should be a floating point value between 0 and 100.\n"+
-			"The upload chunk size is set to its default: %s\n", azureChunkSizeEnvVar, defaultValue))
-		return azureDefaultUploadChunkSize
-	}
-
-	return int(i * humanize.MiByte)
-}
-
 // Azure implements Gateway.
 type Azure struct {
 	host string
@@ -188,6 +155,20 @@ func (g *Azure) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, erro
 		return nil, err
 	}
 
+	azureUploadChunkSize, err = env.GetInt("MINIO_AZURE_CHUNK_SIZE_MB", azureDefaultUploadChunkSizeMB)
+	if err != nil {
+		return nil, err
+	}
+	azureUploadChunkSize *= humanize.MiByte
+	if azureUploadChunkSize <= 0 || azureUploadChunkSize > 100*humanize.MiByte {
+		return nil, fmt.Errorf("MINIO_AZURE_CHUNK_SIZE_MB should be an integer value between 0 and 100")
+	}
+
+	azureUploadConcurrency, err = env.GetInt("MINIO_AZURE_UPLOAD_CONCURRENCY", 4)
+	if err != nil {
+		return nil, err
+	}
+
 	credential, err := azblob.NewSharedKeyCredential(creds.AccessKey, creds.SecretKey)
 	if err != nil {
 		if _, ok := err.(base64.CorruptInputError); ok {
@@ -208,7 +189,9 @@ func (g *Azure) NewGatewayLayer(creds auth.Credentials) (minio.ObjectLayer, erro
 
 	pipeline := azblob.NewPipeline(credential, azblob.PipelineOptions{
 		Retry: azblob.RetryOptions{
-			TryTimeout: azureSdkTimeout,
+			// Azure SDK recommends to set a timeout of 60 seconds per MB of data so we
+			// calculate here the timeout for the configured upload chunck size.
+			TryTimeout: time.Duration(azureUploadChunkSize/humanize.MiByte) * 60 * time.Second,
 		},
 		HTTPSender: pipeline.FactoryFunc(func(next pipeline.Policy, po *pipeline.PolicyOptions) pipeline.PolicyFunc {
 			return func(ctx context.Context, request pipeline.Request) (pipeline.Response, error) {
@@ -914,7 +897,6 @@ func (a *azureObjects) PutObject(ctx context.Context, bucket, object string, r *
 	blobURL := a.client.NewContainerURL(bucket).NewBlockBlobURL(object)
 
 	_, err = azblob.UploadStreamToBlockBlob(ctx, data, blobURL, azblob.UploadStreamToBlockBlobOptions{
-		BufferSize:      azureUploadChunkSize,
 		MaxBuffers:      azureUploadConcurrency,
 		BlobHTTPHeaders: properties,
 		Metadata:        metadata,

--- a/cmd/gateway/azure/gateway-azure_test.go
+++ b/cmd/gateway/azure/gateway-azure_test.go
@@ -20,12 +20,8 @@ import (
 	"encoding/base64"
 	"fmt"
 	"net/http"
-	"os"
 	"reflect"
-	"strconv"
 	"testing"
-
-	"github.com/dustin/go-humanize"
 
 	"github.com/Azure/azure-storage-blob-go/azblob"
 	minio "github.com/minio/minio/cmd"
@@ -267,37 +263,4 @@ func TestCheckAzureUploadID(t *testing.T) {
 			t.Fatalf("%s: expected: <nil>, got: %s", uploadID, err)
 		}
 	}
-}
-
-func TestParsingUploadChunkSize(t *testing.T) {
-	key := "MINIO_AZURE_CHUNK_SIZE_MB"
-	invalidValues := []string{
-		"",
-		"0,3",
-		"100.1",
-		"-1",
-	}
-
-	for i, chunkValue := range invalidValues {
-		os.Setenv(key, chunkValue)
-		result := getUploadChunkSizeFromEnv(key, strconv.Itoa(azureDefaultUploadChunkSize/humanize.MiByte))
-		if result != azureDefaultUploadChunkSize {
-			t.Errorf("Test %d: expected: %d, got: %d", i+1, azureDefaultUploadChunkSize, result)
-		}
-	}
-
-	validValues := []string{
-		"1",
-		"1.25",
-		"50",
-		"99",
-	}
-	for i, chunkValue := range validValues {
-		os.Setenv(key, chunkValue)
-		result := getUploadChunkSizeFromEnv(key, strconv.Itoa(azureDefaultUploadChunkSize/humanize.MiByte))
-		if result == azureDefaultUploadChunkSize {
-			t.Errorf("Test %d: expected: %d, got: %d", i+1, azureDefaultUploadChunkSize, result)
-		}
-	}
-
 }

--- a/pkg/env/env.go
+++ b/pkg/env/env.go
@@ -18,6 +18,7 @@
 package env
 
 import (
+	"strconv"
 	"strings"
 	"sync"
 )
@@ -73,6 +74,16 @@ func Get(key, defaultValue string) string {
 		return v
 	}
 	return defaultValue
+}
+
+// GetInt returns an integer if found in the environment
+// and returns the default value otherwise.
+func GetInt(key string, defaultValue int) (int, error) {
+	v := Get(key, "")
+	if v == "" {
+		return defaultValue, nil
+	}
+	return strconv.Atoi(v)
 }
 
 // List all envs with a given prefix.

